doc/specs/#980 - Apps and Features entries version mapping.md

```diff
@@ -0,0 +1,193 @@
+---
+author: Yao Sun @yao-msft
+created on: 2022-05-09
+last updated: 2022-05-15
+issue id: 980
+---
+
+# Mapping of Apps and Features version and Winget package version
+
+For [#980](https://github.com/microsoft/winget-cli/issues/980)
+
+## Abstract
+
+Some Winget packages may have the concept of marketing version and internal version. So they may have different version values in Winget manifest and in Apps and Features registry entry. Usually, the marketing version will be in the Winget manifest and internal version will be written in the Apps and Features registry.  
+
+In this doc, Winget manifest version will be referred to as Winget version and Apps and Features version in registry will be referred to as ARP version.
+
+## Inspiration
+
+Winget should try to map or correlate the ARP version of an installed package to its Winget version. This will help Winget List/Upgrade better determine if there's an upgrade available.
+
+## Solution Design
+
+### Version parsing and comparison logic in Winget
+
+Versions are parsed by:
+1. Splitting the string based on the split character (`.`)
+2. Parsing a leading, positive integer from each split part
+3. Saving any remaining, non-digits as a supplemental value
+
+Versions are compared by:  
+ for each part in each version  
+  if both sides have no more parts, return equal  
+  else if one side has no more parts, it is less  
+  else if integers not equal, return comparison of integers  
+  else if only one side has a non-empty string part, it is less  
+  else if string parts not equal, return comparison of strings  
+
+For example:  
+Version `1` is less than version `2`  
+Version `1.0.0` is less than version `2.0.0`  
+Version `0.0.1-alpha` is less than version `0.0.2-alpha`  
+Version `0.0.1-beta` is less than version `0.0.2-alpha`  
+Version `0.0.1-alpha` is less than version `0.0.1-beta`  
+Version `0.0.1-alpha` is less than version `0.0.1` 
+Version `13.9.8` is less than version `14.0`  
+Version `1.0` is equal to version `1.0.0`  
+
+Both Winget version and ARP version will use the above parsing and comparison logic.
+
+### How Winget collects and stores ARP version
+
+The `DisplayVersion` field under `AppsAndFeaturesEntries` will be treated as ARP versions of the package.
```

Comment:
What will happen if a single app has multiple AppsAndFeaturesEntries for all of its components (e.g - Firefox) ? Is it expected that the first entry in AppsAndFeaturesEntries should be the "primary" component ARP entry?

Reply:
I'm beginning to wonder if we need to have a `primary` boolean for this, as when I was trying to fix the table this came up too. I don't think we should assume the array is sorted properly, as the current tooling to get AppsAndFeaturesEntries just uses whatever order the table is in (so alphabetical, meaning things like "Microsoft VC Redist 2013-2015 x64" could come before "Mozilla Firefox 100.0").

Reply:
> I'm beginning to wonder if we need to have a `primary` boolean for this

We do, although I would expect that if there are multiple components listed, only the primary one needs to have the `DisplayVersion` set.  Certainly a potential pitfall though.  Having flags like primary (and others) would also enable the complex scenarios that your PR attempted to address (with the proper internal tracking of course).